sandbox/cgroup/monitor.go

```diff
@@ -94,115 +117,131 @@ func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
+				// in this CGroup
+				iw.removePathWithLockHeld(fullPath)
+				watch.folders = append(watch.folders[:i], watch.folders[i+1:]...)
+				break
+			}
+		}
+		if len(watch.folders) == 0 {
+			// and the group can be removed from the watching
+			iw.removeGroupWithLockHeld(watch)
+			// If all the files/folders of this CGroup
+			// have been deleted, notify the client that
+			// it is done.
+			//
+			// XXX: this has to be a go-routine to avoid holding the lock while waiting on dbus
+			go watch.sendClosedNotification()
+		}
+	}
 }
 
-func (iw *inotifyWatcher) removePath(fullPath string) {
+func (iw *inotifyWatcher) removePathWithLockHeld(fullPath string) {
 	iw.pathList[fullPath]--
 	if iw.pathList[fullPath] == 0 {
 		iw.wd.RemoveWatch(fullPath)
 		delete(iw.pathList, fullPath)
 	}
 }
 
-// processDeletedPath checks if the received path corresponds to the passed
-// CGroup, removing it from the list of folders being watched in that CGroup if
-// needed. It returns true if there remain folders to be monitored in that CGroup,
-// or false if all the folders of that CGroup have been deleted.
-func (iw *inotifyWatcher) processDeletedPath(watch *groupToWatch, deletedPath string) bool {
-	for i, fullPath := range watch.folders {
-		if fullPath == deletedPath {
-			// if the folder name is in the list of folders to monitor, decrement the
-			// parent's usage counter, and remove it from the list of folders to watch
-			// in this CGroup (by not adding it to tmpFolders)
-			iw.removePath(fullPath)
-			watch.folders = append(watch.folders[:i], watch.folders[i+1:]...)
-			break
+func (iw *inotifyWatcher) removeGroupWithLockHeld(toRemove *groupToWatch) {
+	for i, grp := range iw.groupList {
+		if grp == toRemove {
+			iw.groupList = append(iw.groupList[:i], iw.groupList[i+1:]...)
 		}
 	}
-	if len(watch.folders) == 0 {
-		// if all the files/folders of this CGroup have been deleted, notify the
-		// client that it is done.
-		watch.sendClosedNotification()
-		return false
-	}
-	return true
 }
 
 func (iw *inotifyWatcher) watcherMainLoop() {
 	for {
 		select {
-		case event := <-iw.wd.Event:
-			if event.Mask&inotify.InDelete == 0 {
+		case err, ok := <-iw.wd.Error:
+			if err != nil {
+				logger.Debugf("failed to read inotify event: %s", err)
+			}
+			if !ok {
+				// the Error channel is closed after Event so there's nothing else to read
+				return
+			}
+		case event, ok := <-iw.wd.Event:
+			if !ok {
+				// the Error channel is closed next so check it for errors before returning
 				continue
 			}
-			var newGroupList []*groupToWatch
-			for _, watch := range iw.groupList {
-				if iw.processDeletedPath(watch, event.Name) {
-					newGroupList = append(newGroupList, watch)
-				}
+			if event.Mask&inotify.InDelete == 0 {
+				continue
 			}
-			iw.groupList = newGroupList
-		case newWatch := <-iw.addWatchChan:
-			iw.addWatch(newWatch)
+			iw.processDeletedPath(event.Name)
```

Comment:
maybe the previous code had problem, but the new code is also strange, having a main go routine that needs to get a mutex feels a bit wrong

Reply:
Indeed, I removed this code change again. I tried to solve the issue that "inotifyWatcher.MonitorDelete()" does not currently return an error if something goes wrong but I agree that the attempt was bad so I removed it again.